Window MRU, now with a UI

rustn00b opened the pull request:
This PR is an updated take on #977 which this PR supercedes. It 
provides a UI to navigate currently open windows sorted in most-recently-used
order (MRU). At present it is not yet ready for prime time, most notably
because features like animations are missing. However (imho) it has proven to
be quite usable!

Because it includes a separate UI to navigate thumbnails of the open windows, it
should not suffer from the same shortcoming as #977 where the latter would mess
up the current disposition of tabs and workspaces because it relied on actual
focus changes to perform the MRU traversal.

Here's the list of what is supported:
- activate the MRU UI using either `Alt-Tab` or `Alt-Grave`, the former traverses
  among all open windows ("None" filter), while the latter only traverses
  windows that share the same `app-id` as the currently focused window ("AppId"
  filter).

  Each window is represented by a thumbnail version of the original.

  Adding `Shift` to the key binding makes the traversal happen in the opposite
  order.
- when the UI is opened:
  - regular key bindings are filtered.
  - using `Alt-Tab` or `Alt-Grave` allow switching.
    traversal modes on the fly. Toggling the `AppId` filter uses the `app-id`
    of the current MRU selection (as opposed to the window that had the focus
    when the UI was opened).
  - releasing the `Alt` key transfers the focus to whatever thumbnail was
    selected in the MRU UI.
    
    Note: This implies that using key bindings that don't use the `Alt` modifier
    won't behave quite as expected because the UI will not close when they
    are released. So for now at least it's `Alt-*` only.
  - `Esc` key closes the MRU UI and returns focus to the
    window that originally held it when the UI was invoked.
  - `q` close the window/applicatoin for the currently selected
    thumbnail.
  - `a`, `o` or `w` changes the "scope" of the windows shown in the MRU UI:
    - `a`: consider all open windows
    - `o`: consider only windows on the current _Output_
    - `w`: consider only windows in the current _Workspace_
  - `j`/`k` and `Left`/`Right` can also be used to navigate thumbnails.
  - `Home`/`End` jump to the first (respectively last) thumbnail.

Key bindings are all **hard coded**. Those that open the MRU UI (`Alt-Tab`,
`Alt-Grave` and their `Shift`-ed variants) use a lower binding priority so
they shouldn't interfere with bindings already set in the configuration file.

Most glaring remaining todos:
- [ ] Animations
- [ ] Support for pointer actions, eg. clicking on thumbnail should close
      the UI and focus the corresponding window.
- [ ] Maybe some others, ymmv.

Despite all that, hope you all enjoy!

jbms commented:
It would be great for all the key bindings to be fully customizable. 

Also it would be great if this could be fully controlled by ipc (e.g. list of window IDs to include in preview), so that you could have layer shell application that handles its own key bindings and then modifies the preview.

rustn00b commented:
>  It would be great for all the key bindings to be fully customizable.

The tricky bit here is that the UI needs to be dismissed with the mod key (`Alt` for now) is released. Hard-coding the bindings side-steps that problem by just wiring the behavior directly in Niri. Something more flexible would involve figuring out what mod (mods?) were used to open the UI to then detect when they are released. Doable, but might be clunky, especially if there is a more general way to deal with key release bindings in the wings.

Wrt bindings used within the UI itself, either the config would need to support some form of "mode" or maybe a separate binding section would be needed for this UI. Idk.

> Also it would be great if this could be fully controlled by ipc.
This shouldn't be too difficult, basically the UI first builds a list of window IDs when it first opens. To fit into what you describe, the only difference would be to allow that list to be fed from IPC.

Although these ideas are definitely interesting, I think there's more work to be done on just the base functionality before tackling more new stuff (at least on my side)!

YaLTeR commented:
Before looking into the code, would it be feasible to have 1 commit with the previous MRU PR (that I reviewed) and then 1 commit with all the new stuff? So it's easier for me to review just the new changes.

> basically the UI first builds a list of window IDs when it first opens

Does it work with windows opening/closing dynamically?

YaLTeR commented:
Oh, if most of the code changed, then I guess that's not necessary. Thanks for the branch, I can use it as a reference, so just keep it up for now.

I see that you bumped MSRV from 1.80 to 1.83 and applied some clippy lints (is_none_or). Is that strictly necessary? I don't have a concrete MSRV policy for niri but if it's just something minor like is_none_or then I'd prefer keeping 1.80 MSRV for now. Regardless, those changes should be in a separate PR.

There's also one formatting change to `window_matches()` that I noticed, why did that happen?

rustn00b commented:
> I see that you bumped MSRV from 1.80 to 1.83 and applied some clippy lints (is_none_or). Is that strictly necessary? 

It was needed to support `ControlFlow::Break` / `ControlFlow::Continue`, which made things look clearer than the version I came up with before.

> Regardless, those changes should be in a separate PR.

So, I'll wait on your opinion regarding the appropriateness of using `ControlFlow` in this PR, and assuming you are in agreement, then I'll break out he MSRV part and do a separate PR.

rustn00b commented:
Hi there!

> May I ask whether an option to disable this feature would be available?

Not a bad idea, one could easily imagine a user either not wanting interfering keybinds or just wanting to use some different external tool to navigate recently used windows. I'll push this onto the todo list.

 > I would personally like to be able to use Super + Tab

Amusingly the PR from which this one is derived _did_ use `Super + Tab`, but the consensus on the comment thread at the time was that `Alt + Tab` was preferable! Anyway, this seems like a pretty good indication that a little time should be spent on making this at least somewhat configurable. One possible step in that direction would be to allow configuration of the modifier that opens the UI, e.g. `<MOD> + Tab`. Also added to the todos.

elcste commented:
This is amazing!!! I tried sway a few years ago but missed a good MRU switcher and went back to Gnome. Alt+grave for same app is and `q` to close is the icing on the cake!

I know this is not complete, but my feedback is that the animation is a little much for me when quickly switching back and forth between two windows. It looks good when pressing tab multiple times to select a window. I think Gnome Shell has a timeout and doesn't show an animation for their switcher if it is invoked very quickly.

Thanks again!

rustn00b commented:
0a7df425 inhibits the thumbnail selection animation unless the MRU UI was open for longer than a threshold duration. The threshold is hard-coded to 250ms, if that settings turns out not to be comfortable for everyone then it could be migrated to the "recent-windows" configuration section. Until such a time, I'm partial to keeping config clutter under control.

elcste commented:
That's a great improvement. I might play with the timeout and try slightly longer. I'll report back if I really prefer it.

One other thing I've noticed: If I press Alt+Tab with only one window open – or Alt+Grave with only one window open of the same app – focus doesn't return correctly to the window. Clicking in the window doesn't help. I have have to do something outside that window, like switch workspaces, open fuzzel or invoke this switcher (like I can Alt+Tab if I got stuck on Alt+Grave). The apps I tried with were Alacritty, Ptyxis, Firefox and Sublime Text in case it matters.

rustn00b commented:
> If I press Alt+Tab with only one window open – or Alt+Grave with only one window open of the same app – focus doesn't return correctly to the window

@elcste, I haven't run into this personally although I do use alacritty as my "test subject". My usual (manual...) tests include the cases you describe (calling the UI with 0, 1, several windows). It's also intriguing that even clicking in the window doesn't trigger a focus update. The only similar situation I've encountered is with niri-in-niri and switching windows in the "outer-niri" while a modifier key was active in the "inner-niri".

Is this something you've encountered with just the latest version from yesterday or is this something you've noticed in older versions?

elcste commented:
I think I noticed it before you added the animation threshold. I can't remember now if you had added the IPC commit or not.

A little more I noticed. It seems like it could relate to the Alt key still being active somehow? I'm seeing things that also happen if I hold the Alt key.
- After using the switcher with one Ptyxis window, when I type I get the app error "visual bell" of invalid input.
- In an app with a menu, the mnemonic letters are underlined.
  - In Sublime Text, the main menu bar. And focus seems to be stuck on the menubar, which I can flick
  - In Ptyxis and Gnome Disks if I press Alt+Grave with the app menus open, I see the mnemonic letters in those menus. 

I use some xkb options that move around some modifier keys like Alt, but I get the same result with the stock config file.

elcste commented:
@rustn00b That must have been it – I'm not seeing it after your latest updates :-)

YaLTeR commented:
Haven't looked at the code yet, some feedback from 2 minutes of playing:
1. Wow this is really cool! With anims and all even
2. Wait you even got the window titles working, nice
3. It should be `H` and `L`, not `J` and `K`
4. Not sure about the `Q`. It's right next to Tab, and there's no confirmation, making it very easy to close a window by accident.
5. The window shouldn't animate from within the Alt-Tab to its position on the workspace. Just fade out or close the menu, no extra anim necessary.
6. The anim when switching between `A`-`O`-`W` modes is a bit jank, the windows move horizontally out of nowhere and other windows jump in.
7. Pressing Alt-Tab shouldn't close the Overview, it should just draw over it.
8. It should be sized relatively to the monitor logical height (like the Overview). Right now, switching from 2560x1440 to 1280x720, or equivalently changing scale from 1 to 2, results in a 2x taller Alt-Tab, when it should result in the same size Alt-Tab.
9. Windows don't redraw in Alt-Tab as can be seen with `weston-simple-egl`, they should.
    - Pay attention to windows that change size like `wleird-resize-loop`. This should work fine.
11. Mouse should be able to at least click to select the target window (and touch and tablet too).

    Right now, mouse seems to go "below" the Alt-Tab. I was actually planning to look into refactoring our Keyboard/PointerTarget to better support modal dialogs, I believe that would be useful for Alt-Tab too to solve this kind of issue.
12. Moving mouse to another monitor switches the Alt-Tab to that monitor. This is fairly abrupt; I think it should stay where it was. Also, all monitors should probably be dimmed, not just the monitor with the Alt-Tab.
12. Not sure how you draw the border in Alt-Tab, but it should definitely be thicker, and also it breaks on rounded corners:

    <img width="66" height="40" alt="image" src="https://github.com/user-attachments/assets/ed9bd99a-c597-46fb-ab2a-6e1a060599f0" />

    I've also seen the bottom border to be missing, which leads me to believe that you draw the border on the original Tile then downscale it, kinda like the Overview. Instead, perhaps try drawing the border on the already-downscaled Tile so that the border itself is not scaled?
13. Design-wise will need to play around and think more about it, remember what GNOME Shell does and what others do. Not prepared to give any feedback rn.

madeleine313 commented:
> It should be H and L, not J and K

I think J and K are fine and work just like they do in e.g. qutebrowser, and I would believe that is the intention.
Is there anything wrong with keeping the current J and K behaviour and adding H and L as well?

YaLTeR commented:
> I was actually planning to look into refactoring our Keyboard/PointerTarget to better support modal dialogs

Scratch that, after looking into it I'm not sure it's worth the trouble, at least for now

rustn00b commented:
@YaLTeR: thanks for the feedback!

I'm not sure how prompt I'll be to get around to all of it though.

> It should be H and L, not J and K

I guess I'll just add H and L and keep J and K for now. (It's fairly trivial to change).

> Not sure about the Q. It's right next to Tab, and there's no confirmation, making it very easy to close a window by accident.

There isn't much finger motion going on there so I have never accidentally closed a window. And I actually have found it quite convenient to just hit Q when I come across a window that has is no longer useful. Technically Mod+Q doesn't ask for a confirmation either. I also got to play with a Mac and their equivalent doesn't prompt either (assuming the app doesn't have it own close confirmation dialog).

> The window shouldn't animate from within the Alt-Tab to its position on the workspace. Just fade out or close the menu, no extra anim necessary.

 Bummer, I kind of really like that (it was also some effort to put together) :/ An older version of this PR was actually implemented without a transition but I found it less intuitive. You would select a thumbnail and the focus would go ... somewhere... With the thumbnail selection animation your eyes track the thumbnail and the direction it leaves towards, so you end up looking where you need to find the newly focused window.

> The anim when switching between A-O-W modes is a bit jank, the windows move horizontally out of nowhere and other windows jump in.

I played around and indeed got into a situation where switching between A-O-W made the selected thumbnail move even though there were enough other thumbnails (for a scope filters) to fill the space to either side. Under those conditions I would expect the selected thumbnail not to move. Back to the drawing board...

> Pressing Alt-Tab shouldn't close the Overview, it should just draw over it.

That's actually a recent change. It's a bit of a cop out because the thumbnail selection (I know you've mentioned it should be removed) didn't have logic to resize the thumbnail to the overview size as opposed to the regular window size. It didn't look very good so I took the easy way out on that one.

> It should be sized relatively to the monitor logical height (like the Overview). Right now, switching from 2560x1440 to 1280x720, or equivalently changing scale from 1 to 2, results in a 2x taller Alt-Tab, when it should result in the same size Alt-Tab.

For now the thumbnails are just scaled to half the logical size of their window counterpart. I've been using it that way for a while with two screens with very different sizes and it's been fine. However I do see situations like using portrait mode with a very tall screen that could lead to things looking weird. I also haven't experimented with changes in scale. I guess I would need to think more about this one.

> Windows don't redraw in Alt-Tab as can be seen with weston-simple-egl, they should.

I only figured out how to use the OffscreenBuffer later in the game, also using a snapshot avoids having to deal with thumbnails resizing on their own. From a usability stand point I haven't been bothered by the fact that the thumbnails don't have the latest content (the use case isn't quite the same as the overview's).

> Mouse should be able to at least click to select the target window (and touch and tablet too).

Agreed, this was already in the todos. And I often click on thumbnails as a reflex and am reminded I have to get around to this one.

> Moving mouse to another monitor switches the Alt-Tab to that monitor. This is fairly abrupt; I think it should stay where it was. Also, all monitors should probably be dimmed, not just the monitor with the Alt-Tab.

Indeed. I wasn't sure what the best approach would be here so I left it the way it was when I started. Dimming all monitors and only displaying the UI on whichever one it was initially opened on shouldn't be too difficult.

> perhaps try drawing the border on the already-downscaled Tile so that the border itself is not scaled?

The focus ring is drawn around the already downscaled thumbnail render, however the border is indeed rendered to a texture which is later downscaled.
For now the radius is static so there is a possible discrepancy between what the thumbnail and the window might be doing.

YaLTeR commented:
> I'm not sure how prompt I'll be to get around to all of it though.

No worries :slightly_smiling_face: 

> There isn't much finger motion going on there so I have never accidentally closed a window. And I actually have found it quite convenient to just hit Q when I come across a window that has is no longer useful. Technically Mod+Q doesn't ask for a confirmation either. I also got to play with a Mac and their equivalent doesn't prompt either (assuming the app doesn't have it own close confirmation dialog).

What key do they have?

Actually we could probably bind Alt+F4 which is a common shortcut, and which seems like it should work nicely with Alt-Tab? Alternatively, maybe hardcode Backspace to close?

> Bummer, I kind of really like that (it was also some effort to put together) :/ An older version of this PR was actually implemented without a transition but I found it less intuitive. You would select a thumbnail and the focus would go ... somewhere... With the thumbnail selection animation your eyes track the thumbnail and the direction it leaves towards, so you end up looking where you need to to find the newly focused window.

Hm, I'll need to play around with it more to see how bad the problem is and how else it can be solved. I guess there should be a simple way to disable this anim for that purpose?

My main problem with the current anim is that it doesn't make spatial sense. Like, your Alt-Tab window previews are clearly overlaid over your window layout, then once you confirm, out of nowhere the original window disappears, and the window preview animates into that position.

> I only figured out how to use the OffscreenBuffer later in the game, also using a snapshot avoids having to deal with thumbnails resizing on their own. From a usability stand point I haven't been bothered by the fact that the thumbnails don't have the latest content (the use case isn't quite the same as the overview's).

Sure it's usable but it also feels incomplete this way, so I'd very much prefer it to live-update. Regarding resizing windows, I'm curious what other Alt-Tabs do with those, like GNOME or KDE.

elcste commented:
@jhasse To each there own :-) – I bind Ctrl+Q to close-window so I can close all windows that way, even apps that don't set that shortcut themselves. (Off topic: You could bind Ctrl+Q in Niri to some no-op behavior so Niri will intercept the shortcut before it gets to any app. I actually bind it to a script that looks at the window title and skips one app I don't want to be able to close.)

But back to this MRU switcher, I love the Q to close feature, but I think it would be great for it to be configurable, either which key or just enable/disable.

I agree with @YaLTeR that the W shortcut seems tricky. I like the idea of the A, O and W filters but feel like maybe they should be all next to each other rather them mnemonics. Like 1, 2 and 3 or maybe even [, ] and \ (varying keyboard layouts could make this hard).

rustn00b commented:
Commit 12ecae6 addresses the following:
- Changed key binding to close a thumbnail to `<Mod>+Shift+q`. So this isn't
  really making it configurable as suggested by @elcste, but should reduce the
  likelihood of accidentally closing an application (because the most frequent
  form of navigation - moving from most recent to least recent - doesn't
  involve the `Shift` key).
- Keep the MRU UI on the display it was initially opened in. The mouse pointer
  is also boxed into that display.
- Thumbnails can be selected using the mouse pointer (and presumably using
  a touch action, but I couldn't test if it actually works).
- Added thumbnail fade-out animation when the MRU UI is closed. This was intended
  but I forgot to include it at some point.
- Fix thumbnail open animation that only had alpha on the Window title but not
  on the thumbnail itself.
- Added `<Mod>+h` and `<Mod>+l` for MRU navigation (without removing `j`/`k`)
- Fix unwanted motion when switching between scopes (point raised by @YaLTeR)

Regarding `A`/`O`/`W`, my thinking is that having a single binding, say
`<Mod>+s` (for scope) that just cycles through the different scopes is
probably more intuitive. A little indication of the current scope would
also be nice I suppose.

As a first step on the road to acceptance that the selection animation is going
to get tossed, I added an `enable-selection-animation` toggle for it in the
`recent-windows` part of the configuration (defaults to `false`).

To address the "not knowing where to look" issue that drove my perceived need
for a thumbnail selection animation, I was thinking that  the `FocusWindow`
action could adjust the view so as to position the selected window in the same
on-screen location as when it was last focused. While this won't really tell you
where to look, it would leverage your memory of where the window _should_ be.
This would involve keeping some extra state in the `ScrollingSpace`.

elcste commented:
> presumably using a touch action, but I couldn't test

I can test touch in the next couple days :-)

elcste commented:
Touch does work to select thumbnails.

I like the reduced animations when switching.

jbms commented:
The key bindings all really ought to be configurable --- I guess the issue is just the lack of a general mechanism for doing that in niri at the moment?

Vortriz commented:
I was testing this out and everything has been working perfectly!

Just one thing that I noticed: the window title displayed under the thumbnail is quite blurry. I am using my display on 1.5 scale rn, not sure if that could be the issue.

YaLTeR commented:
1. I don't want the whole app icon resolution and rendering logic in niri, at least not now. 
2. I have a suspicion that niri users may generally have many terminal windows open at once, making this application grouping less useful than on GNOME.

YaLTeR commented:
Thanks for the response!

> I'm afraid you're making obvious my very limited grasp of how things work on the Smithay end. I'll take the comment to mean I should probably be doing things differently.

This RenderTarget is a niri-specific thing actually, I made it for block-out-from (it's only used for that). Probably not obvious from the generic-sounding name.

> Reverting to 1.80.1 would require replacing a handful of `is_none_or`s.

If that's not a big deal, I'd prefer to do that. Doesn't sound like an important enough thing on its own to warrant bumping the MSRV. (Even if realistically bumping to 1.83 is likely a non-issue).

> I just tab through the first few and read the content directly in the thumbnails (typically shells). I guess there is some margin for that content to still remain readable while making the thumbnails smaller, but not all that much...

Hmm, that is a good point. With a Windows-style menu you don't get to do much reading. Now that I think about it, on Windows, hovering Alt-Tab for a bit will reveal the window, letting you inspect and pick the right one, but I'm not sure this will work well on niri (where it will cause much unnecessary scrolling around). I will try the current MRU design some more then and see how it feels.

> Making the scale configurable, as suggested by @Sempyos, could be an option. Or maybe using the scroll wheel? A more Windows-y layout (spread on multiple rows) would require a more significant rework (I'm not quite ready to take that on just yet!).

Let's hold off on any changes here for now until I got a better feel.

> I unfortunately won't have much spare time in the coming weeks. I'll try to get through the simpler stuff quickly and back out the fixes I was trying to sort out for the rounded corners (which weren't working anyway).

Alright, I'll wait for you to deal with the easy fixes and notify me here, then I'll see if I can finish this PR myself. Good timing before I dive into some other large refactor :)

YaLTeR commented:
@rustn00b is it good to take this over?

YaLTeR commented:
> @rustn00b is it good to take this over?

Well I already got started with my changes to this PR so yeah

rustn00b commented:
@YaLTeR , I have a few minor things still queued up (no longer close overview if MRU is called up and some clean ups). I'll push them today.

There are still some outstanding issues:
- Corner radius handling is incorrect (I was aiming to copy the corner radius set on the window)
- The selection animation still uses an Offscreen render (bcs I haven't figured out how to scale without first going through a texture)
- I just recently noticed there's wonkiness when CSD is used in the selection animation: there's a gap between the title and window.

@Sempyos also pointed out the following:

> 2. There is sometimes a noticeable delay when clicking outside of the MRU to close it which skips the fade out animation

This one I hadn't noticed yet but I must also admit that I don't use the feature much so I'm less likely to come across it.

>  3. The MRU does not close on a tap/touch outside of it

Strange, I thought that was handled. Intended behavior was that when a thumbnail is touched/tapped then that triggers the MRU close with that thumbnail selected. Conversely, if the tap/touch was outside any thumbnail then the MRU is just dismissed.

 > 4. The MRU does not inhibit applications below from receiving inputs while open, which may or may not be an issue
 
 Strange again, I thought inputs (keyboard, pointer and tap) were all intercepted. I'm assuming I missed some other source.

> 6. ..., However, it may be ideal for touchscreen support to be able to drag to scroll thumbnails

Agreed, that one was on the todo list but I haven't gotten around to it.

> 7. Thumbnails have no corner radius.

This was the one I've been trying to address recently, as yet somewhat unsuccessfully.

> 8.  Currently difficult to tell which is the focused window in the picker without a border/focus ring

That's what initially led me to only add the title under the focused thumbnail (though that was removed in 
[ae71349](https://github.com/YaLTeR/niri/pull/1704/commits/ae71349f8d7b63ecc3f79803a8eca575faaa62aa).

Maybe a thicker focus ring is needed, idk.


> is it good to take this over?

@YaLTeR, just to state the obvious: you're more than welcome to plug away at this PR! :) I can't exactly claim that I've been blazingly fast.

YaLTeR commented:
Interestingly, the GNOME Shell Alt-Tab binds Q to close application, and W to close one window (when in the window selector): https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/9d904a804e73c97a1ecde406f395cd77a53f10e7/js/ui/altTab.js#L199

For niri I think it makes sense for Q to close a window given the presumably-common "several terminals" usecase.

yuukibarns commented:
I wonder if it is intended that the mru ui is independent of the `focus-window-previous` action? Window switching via mru ui seems not to update the previous window in `focus-window-previous` action. For example, if the `focus-window-previous` is between window A and B, and we are focusing on window A.  Then when we switch to window C via mru ui and call `focus-window-previous` again, we will come back to window B instead of window A.

yuukibarns commented:
The prompt `S`cope: `A`ll `W`orkspace `O`utput doesn't scale with the `scale` option.

YaLTeR commented:
Updates:
- window title is now also hidden when window is blocked out
- improved visibility somewhat especially when there's text underneath the window title
- urgent windows are now indicated in alt-tab
- pointer selection no longer scrolls the ui
- added WindowFocusTimestampChanged event, now it should be good to make custom MRU switchers or logic

rustn00b commented:
@YaLTeR: Awesome work you've been doing on this, thanks!

#### Scope binding
> question, is the s to cycle scope needed much? Or we can leave just a/w/o. Also: maybe we should rename O - Output to M - Monitor?

Using "monitor" instead of "output" might indeed more appropriate since that's probably what users would expect it to be called (but since the term "output" is also used elsewhere - e.g. `niri msg outputs`, idk). The `s` binding was added in response to the following comment from @elcste:

>  I like the idea of the A, O and W filters but feel like maybe they should be all next to each other rather them mnemonics.

The  shortcut was intended to keep everything in a single location for those who like that better, while still having a somewhat intuitive key binding. I've actually used it occasionally too.

#### Thumbnail scale
> does it work for you? Maybe we can reduce size even further? You can play with the PREVIEW_MAX_HEIGHT constant.

With my monitor it's already on the too-small-to-read side of things, I can still figure out which window is which using the shape of the text but I can no longer just read it. Not exactly a blocking issue, but I found the larger thumbnails more comfortable (though I may be alone in that respect ¯\_(ツ)_/¯ ).

YaLTeR commented:
Oops sorry it's `highlight { corner-radius 32; }` not previews of course

YaLTeR commented:
Default is 0, this was just an example

YaLTeR commented:
Something I've noticed about the bindings. For the filter (Alt-Tab vs. Alt-grave) you want it to always apply on the binding, i.e. pressing Alt-Tab should always switch to no filter, and pressing Alt-grave should always switch to current app's windows filter. However, for the scope (All/Workspace/Output) you don't really want it to always apply on every press, because then you can't switch between scopes while Alt-Tab is open. So maybe the `scope=""` property in binds should rather mean *default* scope (and only apply when initally opening Alt-Tab), and not scope on every key press?

Sempyos commented:
I'm going to attempt to play around with this later and provide my feedback, but from the commits it's been looking very nices ! 
Lovely works

YaLTeR commented:
Updates:
- did most of the internal cleanups
- improved closing fade-out anim (made it offscreen the contents to avoid transparent compositing artifacts)
- changed bind `scope` to affect only initial opening of the switcher, as I mentioned in the comment above
- reimplemented `focus-window-previous` action in terms of the (better) MRU focus timestamp logic
- wrote all docs: https://github.com/YaLTeR/niri/pull/1704/commits/c65916e273d603d7ff15ac5d4f3d7d9bfa3aec11

By now this is very close to merging, I just need to do a final pass over the diff. So if you find any more issues, please write about them.

Vortriz commented:
The docs mention:

> The way this works is by finding all regular binds corresponding to these actions and taking just the trigger key without modifiers.
For example, if you have <kbd>Mod</kbd><kbd>Shift</kbd><kbd>C</kbd> bound to `close-window`, in the window switcher pressing <kbd>C</kbd> on its own will close the window.

What will happen if <kbd>Mod</kbd><kbd>Shift</kbd><kbd>C</kbd> and <kbd>Mod</kbd><kbd>C</kbd> both are binded to permissible actions? Which one will be triggered?

yuukibarns commented:
I seem to find something not quite intuitive: when we switch from window A to an empty workspace below, then mru ui will still put the window A at the first, which means we will switch to window B after the window A in the list instead of more intuitive one, i.e., window A itself. However `focus-window-previous` does switch back to window A and then cycle between it and window B.

stefonarch commented:
Labwc doesn't change order if the switcher is opened on an empty workspace, kwin does a strange thing when not filtering per workspace: 3 windows, switching between 1 and 2 as most recent, when opening the switcher on the empty workspace 3 is selected...

YaLTeR commented:
I additionally checked Windows. There, if you minimize all windows, Alt-Tab keeps the first window selected (doesn't switch next). But Alt-Shift-Tab also keeps the first window selected.

I think it's the most intuitive if in this case Alt-Tab keeps the first window, while Alt-Shift-Tab still selects the very last window. So that's what I just pushed in niri.

Also made scroll wheel binds work in the MRU (they are also in the "automatically picked up from normal binds" category, so make sure they're bound).

YaLTeR commented:
Alright, everything's squashed now, should be good to merge. Unless some last minute issues come up.

Vortriz commented:
Tested with default config. Working without any hitches :+1:

yuukibarns commented:
I have a somewhat nuance feature request for the case when there is no window focused. Currently, the selected window will popup abruptly without any sliding animation but in the normal case there is a sliding animation from the right to the center.

yuukibarns commented:
Sorry for my bad English but I mean that when there is a focused window (the normal case), the switcher will arise as the second window slides in from the right to the center as if the switcher just appears in the middle of the sliding animation. But in the case when there is no focused window, the current behaviour is popping up directly without any animation. I hope there can also be some "slide from right to center" animation for the "no focused window" case.

yuukibarns commented:
Yes, I see but I mean that I wonder if we can add some slide animation even for the first window. I know this is not for everyone (that's why I call it "somewhat nuance"), but I think the current popup behavior of the first window is a little abrupt.

YaLTeR commented:
Alright let's continue on main. Thanks again @rustn00b for all of the work, and everyone for intermediate testing!

rustn00b commented:
Thanks to all of those who chimed in here and made this happen!
